access/common/auth/token_provider.go

```diff
@@ -0,0 +1,221 @@
+		return trace.BadParameter("Ctx must be set")
+	}
+	if c.RetryInterval == 0 {
+		c.RetryInterval = defaultRefreshRetryInterval
+	}
+	if c.TokenBufferInterval == 0 {
+		c.TokenBufferInterval = defaultTokenBufferInterval
+	}
+
+	if c.State == nil {
+		return trace.BadParameter("State must be set")
+	}
+	if c.Refresher == nil {
+		return trace.BadParameter("Refresher must be set")
+	}
+	if c.Clock == nil {
+		c.Clock = clockwork.NewRealClock()
+	}
+	if c.Log == nil {
+		c.Log = logrus.NewEntry(logrus.StandardLogger())
+	}
+	return nil
+}
+
+// RotatedAccessTokenProvider is an implementation of AccessTokenProvider
+// that uses OAuth2 refresh token flow to renew the acess token.
+// The credentials are stored in the given persistent store.
+//
+// To have an up-to-date token, one must run RefreshLoop() in a background goroutine.
+type RotatedAccessTokenProvider struct {
+	ctx                 context.Context
+	retryInterval       time.Duration
+	tokenBufferInterval time.Duration
+	state               state.State
+	refresher           oauth.Refresher
+	clock               clockwork.Clock
+
+	log logrus.FieldLogger
+
+	lock  sync.RWMutex // protects the below fields
+	creds *state.Credentials
+}
+
+// NewRotatedTokenProvider creates a new RotatedAccessTokenProvider from the given config.
+// NewRotatedTokenProvider will return an error if the state does not have existing credentials,
+// meaning they need to be acquired first (e.g. via OAuth2 authorization code flow).
+func NewRotatedTokenProvider(cfg RotatedAccessTokenProviderConfig) (*RotatedAccessTokenProvider, error) {
+	if err := cfg.CheckAndSetDefaults(); err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	provider := &RotatedAccessTokenProvider{
+		ctx:                 cfg.Ctx,
+		retryInterval:       cfg.RetryInterval,
+		tokenBufferInterval: cfg.TokenBufferInterval,
+		state:               cfg.State,
+		refresher:           cfg.Refresher,
+		clock:               cfg.Clock,
+		log:                 cfg.Log,
+	}
+
+	err := provider.init()
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	return provider, nil
+}
+
+func (r *RotatedAccessTokenProvider) init() error {
+	var err error
+
+	r.creds, err = r.state.GetCredentials(r.ctx)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	return nil
+}
+
+// GetAccessToken implements AccessTokenProvider()
+func (r *RotatedAccessTokenProvider) GetAccessToken() (string, error) {
+	r.lock.RLock()
+	defer r.lock.RUnlock()
+	return r.creds.AccessToken, nil
+}
+
+// RefreshLoop runs the credential refresh process.
+func (r *RotatedAccessTokenProvider) RefreshLoop() {
+	r.lock.RLock()
+	creds := r.creds
+	r.lock.RUnlock()
+
+	interval := r.getRefreshInterval(creds)
+
+	timer := r.clock.NewTimer(interval)
+	defer timer.Stop()
+	r.log.Infof("Will attempt token refresh in: %s", interval)
+
+	for {
+		select {
+		case <-r.ctx.Done():
+			r.log.Info("Shutting down")
+			return
+		case <-timer.Chan():
+			creds, _ := r.state.GetCredentials(r.ctx)
```

Comment:
We could log the error here.

I think I omitted the check since it checks `creds != nil` further along the way. This is just an optimistic check whether the creds are already fresh enough, and in case of an error we will still proceed as usual and try to refresh + save new creds.